[libpas] Fix local_allocator stop race

Constellation opened the pull request:
#### b1948459dca702969a89d6e2d0298e5cf6e94805
<pre>
[libpas] Fix local_allocator stop race
<a href="https://bugs.webkit.org/show_bug.cgi?id=242422">https://bugs.webkit.org/show_bug.cgi?id=242422</a>
rdar://90119067

Reviewed by Mark Lam.

pas_local_allocator_stop can be called from the target running thread and the scavenger thread.
And it has the same issue to 484126cfed9edfd1e9e0a937fbbddad27dbfe0e5: if the running thread
is suspended just before setting `is_in_use = true`, we will run stop operation again onto the
already stopped local allocator. This breaks the local allocator, and ends up crashing in pas_local_allocator_commit_if_necessary_impl,
because we could revive already decommited local allocator by partially setting stopped local allocator information.
In this case, the rest of the fields are zero.
This patch also adds comments on how TLC decommit works in various conditions.

* Source/bmalloc/libpas/src/libpas/pas_local_allocator.c:
(pas_local_allocator_stop):
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_commit_if_necessary_impl):
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_scavenger_data.c:
(pas_local_allocator_scavenger_data_commit_if_necessary_slow):

Canonical link: <a href="https://commits.webkit.org/252228@main">https://commits.webkit.org/252228@main</a>
</pre>

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_thread_local_cache_layout.c:
```diff
@@ -71,6 +76,36 @@ pas_allocator_index pas_thread_local_cache_layout_add_node(pas_thread_local_cach
+     *                 Checking this bit says it is not decommited. But the latter part is already decommited.
+     */
+    if (begin_page != end_page) {
+        result = (begin_page - PAS_OFFSETOF(pas_thread_local_cache, local_allocators)) / sizeof(uint64_t);
```
I hacked this, but will clean up later (this patch is draft).

Constellation replied:
Also, I will check whether pas_scavenger can work well with these holes.

Constellation replied:
Ugh, this requires some adjustment in designated heap indexes for x64 since x64 has 4KB page while ARM64 darwin has 16KB page (so, behavior is slightly different). I think for now, we should disable TLC decommit feature.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_thread_local_cache_layout.c:
```diff
@@ -71,6 +76,36 @@ pas_allocator_index pas_thread_local_cache_layout_add_node(pas_thread_local_cach
+    PAS_ASSERT(!did_overflow);
+
+    begin_page = pas_round_up_to_power_of_2(pas_thread_local_cache_offset_of_allocator(result), page_size);
+    end_page = pas_round_up_to_power_of_2(pas_thread_local_cache_offset_of_allocator(pas_thread_local_cache_layout_next_allocator_index - 1), page_size);
```
I would like to ensure that we will never have a local-allocator which exceeds the page size. If it is true, then this is safe.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_thread_local_cache_layout.c:
```diff
@@ -71,6 +71,25 @@ pas_allocator_index pas_thread_local_cache_layout_add_node(pas_thread_local_cach
+     *                [ local allocator ]
+     *                 ^
+     *                 Checking this bit says it is not decommited. But the latter part is already decommited.
+     */
```
I would like to check this a bit more.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator.c:
```diff
@@ -242,6 +242,40 @@ bool pas_local_allocator_stop(
 
+    /* The scavenger thread may race with the client thread on calling pas_local_allocator_stop.
+       The scavenger will first suspend the client thread before checking the is_in_use flag.
+       If is_in_use is set, the scavenger will not call pas_local_view_cache_stop, and move on.
```
/pas_local_view_cache_stop /pas_local_allocator_scavenger_data_stop/

Constellation replied:
Fixed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator.c:
```diff
@@ -242,6 +242,40 @@ bool pas_local_allocator_stop(
+       is_in_use flag.
+
+       If the re-check shows that the local_allocator is already stopped, then pas_local_allocator_stop
+       can return early like first pas_local_allocator_scavenger_data_is_stopped check. This is
```
/like first/like the first/

Constellation replied:
Fixed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -77,10 +77,16 @@ static PAS_ALWAYS_INLINE void pas_local_allocator_commit_if_necessary_impl(pas_l
-    if (PAS_LIKELY(allocator->scavenger_data.kind == pas_local_allocator_allocator_kind))
+    if (PAS_LIKELY(allocator->scavenger_data.kind == pas_local_allocator_allocator_kind)) {
+        /* This is safe. Once is_in_use = true is set, the scavenger does not stop this local allocator.
+         * Whenever the scavenger stops a local allocator, it first check is_in_use is false and then sets scavenger_data.kind to
```
/first check is_in_use/first checks if is_in_use/

Constellation replied:
Fixed.

justinmichaud commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator.c:
```diff
@@ -242,6 +242,40 @@ bool pas_local_allocator_stop(
+       before doing the work to stop the local_allocator.
+
+       However, if the client thread is already in the process of executing pas_local_allocator_stop,
+       gets passed the pas_local_allocator_scavenger_data_is_stopped check, and then, gets suspended by
```
nit: past

Constellation replied:
Ah, it is actually `passed` since "the pas_local_allocator_scavenger_data_is_stopped check" is passed.
Probably we should change it to

"gets the pas_local_allocator_scavenger_data_is_stopped check passed,"

MenloDorian replied:
I think "past" is correct.  I initially wrote it as "pass".  Saam corrected it to "passed" because I wrote "pass".  But what I actually meant is "past" as in the pas_local_allocator_scavenger_data_is_stopped check already occurred.

Constellation replied:
I see. Changed, thanks!

justinmichaud replied:
/gets passed/passes/ would be better maybe?

Constellation replied:
Ah, I changed this to `the past ...` and putted it in merge-queue for now.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -77,10 +77,16 @@ static PAS_ALWAYS_INLINE void pas_local_allocator_commit_if_necessary_impl(pas_l
+    if (PAS_LIKELY(allocator->scavenger_data.kind == pas_local_allocator_allocator_kind)) {
+        /* This is safe. Once is_in_use = true is set, the scavenger does not stop this local allocator.
+         * Whenever the scavenger stops a local allocator, it first check is_in_use is false and then sets scavenger_data.kind to
+         * non pas_local_allocator_allocator_kind. If we set is_in_use = true before and kind is pas_local_allocator_allocator_kind,
+         * it ensures that (1) this allocator is not stopped and (2) this allocator will not be stopped until we put is_in_use = false */
```
Regarding "If we set is_in_use = true before", if we get to this code, then we are guaranteed that `is_in_use` is true (set in `pas_local_allocator_try_allocate`), right?  How about rephrasing this sentence to say, "pas_local_allocator_try_allocate already set is_in_use to true before we get here. If kind is pas_local_allocator_allocator_kind, then we are guaranteed that (1) ... (2) ... until we set is_in_use to false"?

Constellation replied:
Actually, `is_in_use` can be `false`, so why we need to check `allocator->scavenger_data.kind` here. If the local allocator is decommitted, then it can be 0 => false.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_scavenger_data.c:
```diff
@@ -104,6 +104,10 @@ void pas_local_allocator_scavenger_data_commit_if_necessary_slow(
         
         done = false;
         
+        /* Taking scavenger_lock ensures that this page will not be decommited while keeping it.
```
By "keeping it", I understand you to mean the operations of setting data->kind back to pas_local_allocator_allocator_kind or pas_local_allocator_view_cache_kind below.  I suggest rephrasing this as, "will not be decommitted while we change data->kind below to indicate that we're keeping the page."

Constellation replied:
Sounds good. Changed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_scavenger_data.c:
```diff
@@ -104,6 +104,10 @@ void pas_local_allocator_scavenger_data_commit_if_necessary_slow(
         
+        /* Taking scavenger_lock ensures that this page will not be decommited while keeping it.
+         * Since data->kind is set to pas_local_allocator_decommitted_kind before decommit,
+         * if it is not pas_local_allocator_decommitted_kind at this point, this allocator is not
```
/at this point/by this point/

Constellation replied:
Changed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -77,10 +77,16 @@ static PAS_ALWAYS_INLINE void pas_local_allocator_commit_if_necessary_impl(pas_l
        Therefore, we cannot assert that is_in_use is true. But, when we return from here, it must be true. */
-    if (PAS_LIKELY(allocator->scavenger_data.kind == pas_local_allocator_allocator_kind))
+    if (PAS_LIKELY(allocator->scavenger_data.kind == pas_local_allocator_allocator_kind)) {
+        /* This is safe. Once is_in_use = true is set, the scavenger does not stop this local allocator.
```
/does not stop/will not stop/

Constellation replied:
Fixed.

MenloDorian reviewed (approved):
r=me with suggestions.

Constellation reviewed (commented):
Thanks!

Constellation committed:
[libpas] Fix local_allocator stop race
https://bugs.webkit.org/show_bug.cgi?id=242422
rdar://90119067

Reviewed by Mark Lam.

pas_local_allocator_stop can be called from the target running thread and the scavenger thread.
And it has the same issue to 484126cfed9edfd1e9e0a937fbbddad27dbfe0e5: if the running thread
is suspended just before setting `is_in_use = true`, we will run stop operation again onto the
already stopped local allocator. This breaks the local allocator, and ends up crashing in pas_local_allocator_commit_if_necessary_impl,
because we could revive already decommited local allocator by partially setting stopped local allocator information.
In this case, the rest of the fields are zero.
This patch also adds comments on how TLC decommit works in various conditions.

* Source/bmalloc/libpas/src/libpas/pas_local_allocator.c:
(pas_local_allocator_stop):
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_commit_if_necessary_impl):
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_scavenger_data.c:
(pas_local_allocator_scavenger_data_commit_if_necessary_slow):

Canonical link: https://commits.webkit.org/252228@main
